### PR TITLE
Adds a disable env var

### PR DESCRIPTION
**Pull request by camilo**
Some processes that run offline actually don't benefit from semian, for instance processes that do maintenance by just moving a ton of resources around concurrently, they pretty much occupy  all the available tickets by design.

For those cases  I'm providing a way of disabling semian.

### CODE REVIEW

**Comment by camilo** [comment]
@csfrancis @Sirupsen @byroot

**Comment by byroot** [comment]
Hum. I didn't like our previous way of opt-in. But I can't find any argument against an opt-out mechanism.

I guess :+1:

**Comment by byroot** [comment]
> they pretty much occupy all the available tickets by design.

That being said, semian is not just about the semaphore. The resource is also protected by a circuit breaker that can be handy even for such scripts and workers, so that they don't hammer a resource used by the rest of the app.

File: lib/semian.rb [hunk 1]
```diff
@@ -149,10 +153,11 @@ def resources
 require 'semian/protected_resource'
 require 'semian/unprotected_resource'
 require 'semian/platform'
-if Semian.supported_platform?
+if Semian.supported_platform? && !Semian.enabled?
```
Comment (csfrancis) [inline lib/semian.rb]: Shouldn't this be ` && Semian.enabled?`

File: lib/semian.rb [hunk 2]
```diff
@@ -149,10 +153,11 @@ def resources
...
   require 'semian/semian'
 else
   Semian::MAX_TICKETS = 0
-  Semian.logger.info("Semian is not supported on #{RUBY_PLATFORM} - all operations will no-op")
+  Semian.logger.info("Semian is not supported on #{RUBY_PLATFORM} - all operations will no-op") if !Semian.supported_platform?
+  Semian.logger.info("Semian is disabled, is this what you really want? - all operations will no-op") if !Semian.enabled?
```
Comment (Sirupsen) [inline lib/semian.rb]: unless

File: lib/semian.rb (hunk 1 above)
Comment (Sirupsen) [inline lib/semian.rb]: :+1:

**Comment by fw42** [comment]
> That being said, semian is not just about the semaphore. 

Agree. The flag should just disable the semaphore instrumentation, but not prevent all the code from being loaded. That could cause lots of NameError issues and stuff like that in the application code. Opt-in code like non-semaphore circuit breakers should still work.

**Comment by camilo** [comment]
> Agree. The flag should just disable the semaphore instrumentation, but not prevent all the code from being loaded. That could cause lots of NameError issues and stuff like that in the application code.

It should not tho, if that were the case exactly the same would happen on unsupported platforms.

**Comment by camilo** [comment]
>  Opt-in code like non-semaphore circuit breakers should still work.

Also that is kinda bad, if we have an enabled/disabled it should disable all the things or none, if we want that feature control  then we need some config DLS ( :-1: ) or many env vars per feature ( :-1: :-1: :-1: ).

The only thing I'm trying to achieve is to opt-out in situations where semian does not matter (admin scripts and the like), and some features might get in the way, also could be useful for testing as in disable and see that is what we hypothesize would happen without circuits and semaphores actually happens.

**Comment by csfrancis** [comment]
I think I lean towards what Flo is saying here. What @camilo is trying to do is make the semaphore operations no-op, which isn't exactly the same as "disabling Semian" (because yes, Semian is more than just semaphores now). Making semaphores no-op is also what we do in the case where we're running on an unsupported platform.

Perhaps this should be called `SEMIAN_SEMAPHORE_NOOP`, or something else that clearly states that this is only disabling semaphores.

**Comment by camilo** [comment]
I guess unsuported platforms only unsupport semaphores in that case is this what is happening here?

```ruby
if Semian.supported_platform? && Semian.enabled?
  require 'semian/semian'
else
```

Is not obvious by looking at that code that it ONLY disables semaphores given that the name of the gem itself is semian.

**Comment by byroot** [comment]
Ok, I'd say go for it. No semaphore will be acquired, but the circuit breaker will still be there so the behaviour shouldn't change overall.

Please cleanup the `if ! / unless` though.

**Comment by csfrancis** [comment]
> I guess unsuported platforms only unsupport semaphores in that case is this what is happening here?

Yeah, that's what it is. Maybe clean this up so the check is something like `sysv_semaphores_supported?` instead of `supported_platform?`.

**Comment by camilo** [comment]
`if ! / unless` does it even matter? aimma just make if `if !!!`

**Comment by byroot** [comment]
Good joke

File: lib/semian.rb [hunk 3]
```diff
@@ -149,10 +153,11 @@ def resources
 require 'semian/protected_resource'
 require 'semian/unprotected_resource'
 require 'semian/platform'
-if Semian.supported_platform?
+if Semian.supported_platform? && Semian.enabled?
   require 'semian/semian'
 else
   Semian::MAX_TICKETS = 0
-  Semian.logger.info("Semian is not supported on #{RUBY_PLATFORM} - all operations will no-op")
+  Semian.logger.info("Semian is not supported on #{RUBY_PLATFORM} - all operations will no-op") if !Semian.supported_platform?
```
Comment (fw42) [inline lib/semian.rb]: :trollface:

**Comment by camilo** [comment]
ok now I added  `if !!!!!!!!!` and changed the names both for the env var and method.

**Comment by camilo** [comment]
hmrr iamma change `enabled?` too to `semaphores_enabled?`

**Comment by camilo** [comment]
iamma merge unless someone objects.

**Comment by csfrancis** [comment]
LGTM.

**Comment by byroot** [comment]
Sure merge away. I have an open PR to update semian. I'll update it once you merge.

**Pull request merged**